File: modules/signal-protocol/android/src/main/java/expo/modules/signalprotocol/KeyGeneration.kt
package expo.modules.signalprotocol

import org.signal.libsignal.protocol.IdentityKeyPair
import org.signal.libsignal.protocol.ecc.ECKeyPair
import org.signal.libsignal.protocol.kem.KEMKeyPair
import org.signal.libsignal.protocol.kem.KEMKeyType
import org.signal.libsignal.protocol.state.KyberPreKeyRecord
import org.signal.libsignal.protocol.state.PreKeyRecord
import org.signal.libsignal.protocol.state.SignedPreKeyRecord
import java.security.SecureRandom

object KeyGeneration {
    const val MAX_VAL: Int = 16777215
    private const val MAX_PRE_KEYS = 100
    private const val REGISTRATION_ID_MASK: Int = 16383
    private val secureRandom = SecureRandom()

    private fun generateIdentityKeyPair(): IdentityKeyPair {
        return IdentityKeyPair.generate()
    }

    private fun generateRegistrationId(): Int {
        return secureRandom.nextInt(REGISTRATION_ID_MASK)
    }

    fun generateSignedPreKey(identityKeyPair: IdentityKeyPair, signedPreKeyId: Int): SignedPreKeyRecord {
        val keyPair = ECKeyPair.generate()
        val signature = identityKeyPair.privateKey.calculateSignature(keyPair.publicKey.serialize())
        return SignedPreKeyRecord(
            signedPreKeyId,
            System.currentTimeMillis(),
            keyPair,
            signature
        )
    }

    fun generatePreKeys(start: Int, count: Int): List<PreKeyRecord> {
        val results = mutableListOf<PreKeyRecord>()
        for (i in 0 until count) {
            val id = ((start + i) % (MAX_VAL - 1)) + 1
            val keyPair = ECKeyPair.generate()
            results.add(PreKeyRecord(id, keyPair))
        }
        return results
    }

    fun generateKyberPreKeys(start: Int, count: Int, identityKeyPair: IdentityKeyPair): List<KyberPreKeyRecord> {
        val results = mutableListOf<KyberPreKeyRecord>()
        for (i in 0 until count) {
            val id = ((start + i) % (MAX_VAL - 1)) + 1
            val kemKeyPair = KEMKeyPair.generate(KEMKeyType.KYBER_1024)
            val signature = identityKeyPair.privateKey.calculateSignature(kemKeyPair.publicKey.serialize())
            val timestamp = System.currentTimeMillis()
            results.add(KyberPreKeyRecord(id, timestamp, kemKeyPair, signature))
        }
        return results
    }

    fun generateKeys(): Registration {
        val identityKeyPair = generateIdentityKeyPair()
        val registrationId = generateRegistrationId()
        val signedPreKeyId = secureRandom.nextInt(MAX_VAL - 2) + 1
        val signedPreKey = generateSignedPreKey(identityKeyPair, signedPreKeyId)
        val start = secureRandom.nextInt(MAX_VAL - MAX_PRE_KEYS - 2) + 1
        val preKeys = generatePreKeys(start, MAX_PRE_KEYS)
        val kyberPreKeys = generateKyberPreKeys(start, MAX_PRE_KEYS, identityKeyPair)

        return Registration(
            identityKeyPair = identityKeyPair,
            registrationId = registrationId,
            preKeys = preKeys,
            kyberPreKeys = kyberPreKeys,
            signedPreKeyRecord = signedPreKey
        )
    }
}

File: modules/signal-protocol/ios/KeyGeneration.swift
//
//  KeyGeneration.swift
//  Pods
//
//  TilliT Native — Signal Protocol bindings
//


import Foundation
import LibSignalClient

struct KeyGeneration {
    static let maxVal: UInt32 = 16777215
    static let maxPreKeys = 100
    static let registrationIdMask: UInt32 = 0x3FFF

    static func generateIdentityKeyPair() -> IdentityKeyPair {
        return IdentityKeyPair.generate()
    }

    static func generateRegistrationId() -> UInt32 {
        return UInt32.random(in: 0...registrationIdMask)
    }

    static func generateSignedPreKey(identityKeyPair: IdentityKeyPair, signedPreKeyId: UInt32) throws -> SignedPreKeyRecord {
        let privateKey = PrivateKey.generate()
        let signature = identityKeyPair.privateKey.generateSignature(message: privateKey.publicKey.serialize())
        return try SignedPreKeyRecord(
            id: signedPreKeyId,
            timestamp: UInt64(Date().timeIntervalSince1970),
            privateKey: privateKey,
            signature: signature
        )
    }

    static func generatePreKeys(start: UInt32, count: UInt32) throws -> [PreKeyRecord] {
        var results = [PreKeyRecord]()
        for i in 0..<count {
            let id = ((start + i) % (maxVal - 1)) + 1
            let privateKey = PrivateKey.generate()
            let publicKey = privateKey.publicKey
            results.append(try PreKeyRecord(id: id, publicKey: publicKey, privateKey: privateKey))
        }
        return results
    }
    
    static func generateKyberPreKeys(startId: UInt32, count: UInt32, identityKeyPair: IdentityKeyPair) throws -> [KyberPreKeyRecord] {
        var kyberPreKeys = [KyberPreKeyRecord]()
        for i in 0..<count {
            let id = ((startId + i) % (maxVal - 1)) + 1
            let keyPair = KEMKeyPair.generate()
            let publicKey = keyPair.publicKey
            let signature = try signKyberPreKey(
                kyberPublicKey: publicKey,
                identityKeyPair: identityKeyPair
            )
            let timestamp = UInt64(Date().timeIntervalSince1970)
            let kyberPreKey = try KyberPreKeyRecord(
                id: id,
                timestamp: timestamp,
                keyPair: keyPair,
                signature: signature
            )
            kyberPreKeys.append(kyberPreKey)
        }
        return kyberPreKeys
    }
    
    static func signKyberPreKey(kyberPublicKey: KEMPublicKey, identityKeyPair: IdentityKeyPair) throws -> Data {
        let message = kyberPublicKey.serialize()
        return identityKeyPair.privateKey.generateSignature(message: message)
    }

    static func generateKeys(existingIdentity: IdentityKeyPair? = nil) throws -> Registration {
        // Multi-device pairing: when an existing identity is provided (linked
        // device import), reuse it instead of generating a fresh one. All the
        // other keys (registration id, signed pre-key, pre-keys, kyber
        // pre-keys) are still generated fresh per-device — only the user-level
        // identity is shared across devices of the same user.
        let identityKeyPair = existingIdentity ?? generateIdentityKeyPair()
        let registrationId = generateRegistrationId()
        let signedPreKeyId = UInt32.random(in: 1...maxVal - 1)
        let signedPreKey = try generateSignedPreKey(identityKeyPair: identityKeyPair, signedPreKeyId: signedPreKeyId)
        let start = UInt32.random(in: 1...maxVal - UInt32(maxPreKeys) - 1)
        let preKeys = try generatePreKeys(start: start, count: UInt32(maxPreKeys))
        let kyberPreKeys = try generateKyberPreKeys(startId: start + UInt32(maxPreKeys), count: UInt32(maxPreKeys), identityKeyPair: identityKeyPair)

        return Registration(
            identityKeyPair: identityKeyPair,
            registrationId: registrationId,
            preKeys: preKeys,
            signedPreKeyRecord: signedPreKey,
            kyberPreKeys: kyberPreKeys
        )
    }
}
